README.md
[![New Relic Experimental header](https://github.com/newrelic/open-source-office/raw/master/examples/categories/images/Experimental.png)](https://github.com/newrelic/open-source-office/blob/master/examples/categories/index.md#new-relic-experimental)

# New Relic One Wall Status Board (nr1-wall-status-board)

![CI](https://github.com/newrelic-experimental/nr1-wall-status-board/workflows/CI/badge.svg) ![GitHub release (latest SemVer including pre-releases)](https://img.shields.io/github/v/release/newrelic-experimental/nr1-wall-status-board?include_prereleases&sort=semver) [![Snyk](https://snyk.io/test/github/newrelic-experimental/nr1-wall-status-board/badge.svg)](https://snyk.io/test/github/newrelic-experimental/nr1-wall-status-board)

This nerdlet is designed to allow key metrics to be observed in a way that makes incidents and platform state easy to recognise and evaulate with little cognitive effort. The panels change colour based on defined thresholds and the historical state is illustrated to provide context. The nerdlet is intended to be displayed full screen on a TV on the wall and be understood at a distance -  but you can of course use it how you like.

## Features
- Fully configurable through UI (which is persistently stored by account)
- Use any NRQL query to drive widgets
- Cross account querying
- Multiple groups of widgets (in 1, 2 or 3 columns)
- Multi-page (including automated cycling through pages!)
- TV Mode
- Numeric and regex thresholds deriving warning and critical states
- Sub query support to further decorate data (extensible through code)
- CSS can be edited to style to your brand (including custom fonts and custom colours)

![Screenshot](gfx/screenshot.png)

## Installation
 - Download repo
 - run `npm install` 
 - run `nr1 nerdpack:uuid -gf`
 - update the [config.json](/nerdlets/status-board-nerdlet/config.json) with your New Relic account ID. 
 - Serve locally `nr1 nerdpack:serve` or deploy using the `nr1` CLI.


## Configuration
All configuration is driven through the application, click the Configuration button and a form will appear below. Each field is described. Some of the more interesting settings are discussed here. Note that for each panel there may be options for the type of panel. Current values include "NRQL query with threshold" which should normally e used and "Place holder". The place holder is simply to allow you to demo the wall board with placeholder panels.

#### Default refresh rate
Each panel will auto refresh at this rate.

#### Auto rotate pages
If there is more than one page of widgets, then if this has a value > 0 then when in TV mode the pages will be cycled in turn.

#### Groups: Number of groups per row
Panels are displayed in groups. This setting allows you to determine if each group is 1 (6 panels per row), 2 (3 panels per row) or 3 (2 panels per row) columns wide.

#### Panels: Field name
When you provide your NRQL query try to alias the value you want to dusplay. This name is then used in the Field name ocnfiguration. This tells the nerdlet which field in the query to display and test against. Sometimes the field name is not what you might expect. If you have trouble enable debug mode on the panel and explore the results displayed in the console. This will show the data returned and importantly the field name applied to the value in the query. For instance in the query `SELECT apdex(duration,0.5) from Transaction` the field name is `score`.

#### Panels: Thresholds
You can provide thresholds for warning and critical states. These can either be numeric thresholds or regular expressions. You can choosew to display the value or provide an label for each state. This might be useful if the value itself is not very meaningful and a label would provide simpler cognitive effort.

#### Panels: Custom Features
An example custom feature is included called `example`. What this demonstrates is how to specify additional NRQL queries for the panel to enrich it further. For instance the demo data shows the numer of transactions for all apps, and then as an enrichment shows how many apps are reporting. Its possible to easily add extra features like this without the need to worry about the loading of data..

#### Panels: Nested Aggregation support
If your NRQL uses nested aggreation you must indicate where the timeseries keyword should be injected. Add the token `[[TIMESERIES]]` in all the applicable places. e.g. 
`select sum(foo) as total from (SELECT count(*) as 'foo' from Transaction facet appName [[TIMESERIES]]) [[TIMESERIES]]`

docs/CHANGELOG.md
## [1.1.1](https://github.com/newrelic-experimental/nr1-wall-status-board/compare/v1.1.0...v1.1.1) (2020-11-27)


### Bug Fixes

* default config bug ([94104dd](https://github.com/newrelic-experimental/nr1-wall-status-board/commit/94104ddbc56db48451732d9ce5a83a12a3b6190b))

# [1.1.0](https://github.com/newrelic-experimental/nr1-wall-status-board/compare/v1.0.0...v1.1.0) (2020-10-30)


### Features

* Nested Aggregation Support ([b095478](https://github.com/newrelic-experimental/nr1-wall-status-board/commit/b095478f3928993c9827dfdd04d5587d2f27b920))

# 1.0.0 (2020-06-08)


### Features

* Update CI workflows and standard OSS files ([#7](https://github.com/newrelic-experimental/nr1-wall-status-board/issues/7)) ([abbff86](https://github.com/newrelic-experimental/nr1-wall-status-board/commit/abbff869ca3d1c65dd5aa215c976a73cfa6f57ba))


components/Configurator/index.js
import React from 'react';
import PropTypes from 'prop-types';
import Form from "react-jsonschema-form";
import MyBaseInput from './Customisations/BaseInput.js'
import { Spinner, AccountStorageQuery, AccountStorageMutation, Toast, Icon, Modal, HeadingText, BlockText} from 'nr1'
import ToolPanel from './Components/ToolPanel'

export default class Configurator extends React.Component {

    static propTypes = {
        schema: PropTypes.object.isRequired,
        dataChangeHandler: PropTypes.func.isRequired
    }

    constructor(props) {
        super(props);
        this.state = { persistentData:null, configVisible: false}
        this.loadFromStorage = this.loadFromStorage.bind(this);
        this.saveToStorage = this.saveToStorage.bind(this);
        this.saveJSONdata = this.saveJSONdata.bind(this)
    }

    async componentDidMount() {
        this.loadFromStorage()
    }

    // static myMethod() {
    //     console.log("foo")
    // }



    saveToStorage(data) {
    /*
    We save one document with meta dat indicating how many other documents there are to load. 
    The data is then saved to multiple documents of 60k size each 
    */

        const chunkSize= 50000 //50k (broke at 60, not sure why)

        const splitSlice = function(str, len) {
            var ret = [ ];
            for (var offset = 0, strLen = str.length; offset < strLen; offset += len) {
              ret.push(str.slice(offset, len + offset));
            }
            return ret;
          }
        let dataStr=JSON.stringify(data)
        let dataChunks=splitSlice(JSON.stringify(dataStr),chunkSize)
        console.log(`Saving data in ${dataChunks.length} ${chunkSize/1000}k chunks`)
        let metaData={chunks:dataChunks.length, totalSize:dataStr.length}
        Toast.showToast({
            type: Toast.TYPE.NORMAL,
            title: 'Saving data',
            description: `Saving config data using x${dataChunks.length} ${chunkSize/1000}k storage documents`,
            sticky: false
        });


        let validDocs=[`${this.props.documentId}-metadata`]
        dataChunks.forEach((chunk,idx)=>{
            validDocs.push(`${this.props.documentId}-chunk-${idx}`)
        })

        AccountStorageMutation.mutate({
            accountId: this.props.accountId,
            actionType: AccountStorageMutation.ACTION_TYPE.WRITE_DOCUMENT,
            collection: this.props.storageCollectionId,
            documentId: `${this.props.documentId}-metadata`,
            document: JSON.stringify(metaData)
        })

        dataChunks.forEach((chunk,idx)=>{
            AccountStorageMutation.mutate({
                accountId: this.props.accountId,
                actionType: AccountStorageMutation.ACTION_TYPE.WRITE_DOCUMENT,
                collection: this.props.storageCollectionId,
                documentId: `${this.props.documentId}-chunk-${idx}`,
                document: JSON.stringify(chunk)
            })
        })
        this.setState({persistentData: data, dataLoaded: true, configVisible: false})
        this.props.dataChangeHandler(data)

        //delete unused chunks, these might be left over from when the config was larger
        AccountStorageQuery.query({
            accountId: this.props.accountId,
            collection: this.props.storageCollectionId,
            fetchPolicyType: AccountStorageQuery.FETCH_POLICY_TYPE.NO_CACHE
        }).then(({ data }) => {
            data.forEach((doc)=>{
                if(!validDocs.includes(doc.id) && doc.id.includes(this.props.documentId)) {
                    console.log(`Deleting unnecessary local storage chunk document ${doc.id}`)
                    AccountStorageMutation.mutate({
                        accountId: this.props.accountId,
                        actionType: AccountStorageMutation.ACTION_TYPE.DELETE_DOCUMENT,
                        collection: this.props.storageCollectionId,
                        documentId: doc.id,
                    })
                }
            })
        })

    }

    loadFromStorage() {

        AccountStorageQuery.query({
            accountId: this.props.accountId,
            collection: this.props.storageCollectionId
        }).then(({ data }) => {
            let metaData=data.find((e) => e.id===`${this.props.documentId}-metadata`)
            if(metaData) {
                console.log(`Loading ${metaData.document.chunks} chunks`)
                let promises=[]
                let results=[]
                for(let i=0; i<metaData.document.chunks; i++ ){
                    promises[i]=AccountStorageQuery.query({
                        accountId: this.props.accountId,
                        collection: this.props.storageCollectionId,
                        documentId: `${this.props.documentId}-chunk-${i}`,
                    }).then(({data})=>{
                        results[i]=data
                    })
                }
                let that=this
                Promise.all(promises).then(function() {
                    let resultStr=""
                    results.forEach((result)=>{
                        resultStr+=result
                    })
                   
                    let parsedString=JSON.parse(resultStr)
                    if(parsedString.length != metaData.document.totalSize) {
                        alert('Config data loaded but unexpected size, it might be corrupt')
                    }
                    that.setState({persistentData: JSON.parse(parsedString), dataLoaded: true})
                    that.props.dataChangeHandler(JSON.parse(parsedString))
                  });

            } else {
                const hasDefault = this.props.default;
                Toast.showToast({
                    type: hasDefault ? Toast.TYPE.NORMAL : Toast.TYPE.CRITICAL,
                    title: "No config data",
                    description: "No previously saved config data was found, initialising with " + hasDefault ? "default" : "empty" + " config.",
                    sticky: false
                });
                if (hasDefault) {
                    this.setState({ persistentData: this.props.default, dataLoaded: true });
                    this.props.dataChangeHandler(this.props.default);
                    //this.saveToStorage(this.props.default); //this can be buggy, instead lets allow user to choose to save.
                }
                else {
                    this.setState({ persistentData: {}, dataLoaded: true });
                }
            }

        });

        // AccountStorageQuery.query({
        //     accountId: this.props.accountId,
        //     collection: this.props.storageCollectionId,
        //     documentId: this.props.documentId,
        // }).then(({ data }) => {
        //   console.log("Config data loaded",data)
        //   this.setState({persistentData: data, dataLoaded: true})
        // });
    }


    saveJSONdata(data) {
        try {
            console.log(data)
            let parsedData=JSON.parse(data)
            this.saveToStorage(parsedData)
            Toast.showToast({
                type: Toast.TYPE.NORMAL,
                title: 'Success',
                description: 'The data was successfully imported.',
                sticky: false
            });
        } catch(e) {
            console.error(e)

           Toast.showToast({
                    type: Toast.TYPE.CRITICAL,
                    title: 'Data Error',
                    description: 'The data provided doesnt parse as valid JSON. Rectifiy and try again.',
                    sticky: true
                });
        }
    }

    render() {

        const { schema, dataChangeHandler, documentId, storageCollectionId, buttonTitle, modalTitle, modalHelp, uiSchema } = this.props  
        const { persistentData,dataLoaded, configVisible } = this.state  

        const log = (type) => console.log.bind(console, type)

        let returnVal, configPanel
        if(configVisible) {
            configPanel=<div><Spinner inline />Loading configuration...</div>
            if(dataLoaded) {
    
                const widgets = {
                    BaseInput: MyBaseInput //a hack to allow us to remove the nr styles on text inputs
                };
    
                    
                const onSubmit = ({formData}, e) => {
                    console.log("Data submitted: ",  formData, JSON.stringify(formData))
                    dataChangeHandler(formData)
                    this.saveToStorage(formData)
                }
    
                const onChange = ({formData}, e) => {
                    this.setState({persistentData: formData})
                }
    
           
                let form = <Form schema={schema} 
                        uiSchema={uiSchema} 
                        widgets={widgets} 
                        formData={persistentData} 
                        onChange={onChange}
                        onSubmit={onSubmit}
                        onError={log("errors")}
                        >
                        <button className="u-unstyledButton btn btn-info" type="submit"><Icon spacingType={[Icon.SPACING_TYPE.SMALL]} type={Icon.TYPE.HARDWARE_AND_SOFTWARE__SOFTWARE__UPSTREAM_CONNECTION} inline />Save</button></Form>
    
                returnVal=<>                    
                    <div className="ConfiguratorForm">{form}</div>
                    <ToolPanel previousData="test" currentData={persistentData} saveHandler={this.saveJSONdata} documentId={documentId} storageCollectionId={storageCollectionId}/>
                    
                </>
                configPanel=<div className="configContainer" hidden={!configVisible} onClose={()=>{this.setState({configVisible: false})}} >
                <HeadingText>{modalTitle ? modalTitle : "Configuration Editor"}</HeadingText>
                <BlockText>{modalHelp ? modalHelp : "Use the form below to configure the application."}</BlockText>
                           <div style={{"marginTop":"2em"}}>{returnVal}</div>
                         </div>
            }
            
    
        } 

         return <div>
            <button className="btn u-unstyledButton configButton" onClick={()=>{this.setState({configVisible: !configVisible})}}>
                <Icon spacingType={[Icon.SPACING_TYPE.SMALL]} type={Icon.TYPE.INTERFACE__OPERATIONS__CONFIGURE} inline ></Icon> {buttonTitle ? buttonTitle : "Configuration"}
            </button>


            {/* <Modal hidden={!configVisible} onClose={()=>{this.setState({configVisible: false})}} >
    <HeadingText>{modalTitle ? modalTitle : "Configuration Editor"}</HeadingText>
    <BlockText>{modalHelp ? modalHelp : "Use the form below to configure the application."}</BlockText>
               <div style={{"marginTop":"2em"}}>{returnVal}</div>
             </Modal> */}

             {configPanel}

        </div>
    }
}


components/BoardElements/WidgetNRQL.js
import React, { Component,  } from 'react';
import PropTypes from 'prop-types';
import { StatusBlock } from './'
import {NerdGraphQuery} from 'nr1'
import  moment from 'moment'



export default class WidgetNRQL extends Component {
    static propTypes = {
        accountId: PropTypes.number.isRequired,
        autoRefresh: PropTypes.number.isRequired,
    }

    constructor(props) {
        super(props);
        this.state = {data: null}
    }

    async componentDidMount() {
        this.loadData()
        this.autoRefresh = setInterval(() => this.loadData(), this.props.autoRefresh ? this.props.autoRefresh*1000 : 60*1000) //auto refresh 1 minute;
    }

    componentDidUpdate(prevProps) {
        if (prevProps.pageRef!==this.props.pageRef){
            this.setState({data: null})
            this.loadData()

        }
    }

    componentWillUnmount() {
        clearInterval(this.autoRefresh);
    }

    loadData() {
        const { config } = this.props
        const { nrql,field, subField, bucketSize, untilSeconds, additionalQueries} = config
        let accountId = config.accountId ? config.accountId : this.props.accountId

        //The history buckets need to be fixed time periods so that the data within them doesnt shift as time progresses. So lets find the end of the last bucket rounded to a fixed time interval such as 5 minutes
        const date_round = function(date, duration) { return moment(Math.floor((+date)/(+duration)) * (+duration)) }
        let now = moment()
        now.subtract(1,'minutes') //exclude the last minute of data
        let endTime = date_round(now, moment.duration(bucketSize, 'minutes'))
        let startTime= endTime.clone().subtract(bucketSize*24,'minutes')
        let sinceAdjusted = Number(bucketSize) +  Math.round(Number(untilSeconds)/60)

        const variables = {
            id: Number(accountId)
        }

        //add any additional queries to the graphql
        let extraNRQL=""
        if(additionalQueries) {
            Object.keys(additionalQueries).forEach((key)=>{
                //only add the recent restriction if no since clause already
                let q=additionalQueries[key].toLowerCase().includes("since")  ? additionalQueries[key] : `${additionalQueries[key]}  since ${sinceAdjusted} minutes ago until ${untilSeconds} seconds ago`
                extraNRQL+=`
                    ${key}: nrql(query: "${q}") {results}
                `
            })
        }

        //Determine where to place TIMESERIES keyword. For Nested aggregation queries we need to place in multiple places so we look for the token [[TIMESERIES]] and replace it
        let recentQuery=nrql.includes('[[TIMESERIES]]') ?
                `${nrql.replaceAll('[[TIMESERIES]]','')} since ${sinceAdjusted} minutes ago until ${untilSeconds} seconds ago`
            :
                `${nrql} since ${sinceAdjusted} minutes ago until ${untilSeconds} seconds ago`;


        let bucketQuery=nrql.includes('[[TIMESERIES]]') ?
                `${nrql.replaceAll('[[TIMESERIES]]', `timeseries ${bucketSize} minutes`)} since ${startTime.unix()} until ${endTime.unix()}`
            :
                `${nrql} since ${startTime.unix()} until ${endTime.unix()}  timeseries ${bucketSize} minutes`;
            
        let query = `
        query($id: Int!) {
            actor {
                account(id: $id) {
                    recent: nrql(query: "${recentQuery}") {results}
                    buckets: nrql(query: "${bucketQuery}") {results}
                    ${extraNRQL}
                }
            }
        }
        `

        const x = NerdGraphQuery.query({ query: query, variables: variables, fetchPolicyType: NerdGraphQuery.FETCH_POLICY_TYPE.NO_CACHE });
        x.then(results => {
            if(config.debugMode===true) {
                console.log(`DEBUG MODE ENABLED: ${config.title}`,results.data.actor.account)
                console.log(`GQL Query was:`,query)
            }

            let bucketData=results.data.actor.account.buckets.results

            let itemCurrentData=null
            let checkedField = field
            let checkedSubField = subField

            if (results.data.actor.account.recent.results.length > 0){
                checkedField = checkFieldName(field, results.data.actor.account.recent.results[0], config.debugMode)

                if (results.data.actor.account.recent.results[0][checkedField] != null) {
                    itemCurrentData = results.data.actor.account.recent.results[0][checkedField]
                } else if (results.data.actor.account.recent.results[0][checkedField] === null) {
                    console.error(`No data found for panel '${config.title}': Please check the NRQL query to ensure it returns a result.`, results.data.actor.account.buckets.results)
                }
                 else {
                    console.error(`Error with '${config.title}' panel: Please supply a field name to access the data returned.`, results.data.actor.account.buckets.results)
                }

                if (results.data.actor.account.recent.results[0][checkedField] != null && typeof results.data.actor.account.recent.results[0][checkedField] === 'object') {
                    checkedSubField = checkFieldName(subField, results.data.actor.account.recent.results[0][checkedField], config.debugMode)

                    itemCurrentData = results.data.actor.account.recent.results[0][checkedField][checkedSubField]

                    if (itemCurrentData === undefined) {
                        itemCurrentData = null
                        console.error(`Error with '${config.title}' panel: The provided sub field name does is incorrect.`, results.data.actor.account.buckets.results)
                    }
                }
            }

            let data = {
                "current": itemCurrentData,
                "history": bucketData.map((item)=>{ return { value: checkedSubField ? item[checkedField][checkedSubField] : item[checkedField], startTime:item.beginTimeSeconds, endTime: item.endTimeSeconds }})
            }

            if(additionalQueries) {
                data.additional={}
                Object.keys(additionalQueries).forEach((key)=>{
                    data.additional[key]=results.data.actor.account[key].results
                })
            }
            this.setState({ data: data  })
        }).catch((error) => { console.log(error); })
    }

    //custom features example
    /*
        This simply displays the results of another query (as specified in the config with the name 'exampleQuery')
        The idea is that the query is already loaded
    */
    featureEXAMPLE(config,data) {
        if(data.additional && data.additional.exampleQuery) { //the query key in the config set to "exampleQuery"
            return { value: `${data.additional.exampleQuery[0].apps} apps reporting`, tooltip: `There could be even more detail here`}
        }
    }

    render() {
        let {config} = this.props
        let {title, roundTo, valueLabel, valueSuffix, thresholdType, thresholdDirection, thresholdCritical, thresholdCriticalLabel, thresholdWarning, thresholdWarningLabel, thresholdNormalLabel, customFeature, link} = config
        let {data} = this.state

        const determineStatus = (val) => {
            let returnType="N"
            if(thresholdType) {
                if(thresholdType == "numeric") {
                    if(thresholdDirection && thresholdDirection=='below') {
                        returnType =  (val <= thresholdWarning) ? "W" : returnType
                        returnType =  (val <= thresholdCritical) ? "C" : returnType
                    } else {
                        returnType =  (val >= thresholdWarning) ? "W" : returnType
                        returnType =  (val >= thresholdCritical) ? "C" : returnType
                    }
                }
                if(thresholdType == "string") {
                    try {
                        if(thresholdWarning) {
                            let regexW = new RegExp(thresholdWarning)

                            if( regexW.test(val) ) {
                                returnType= "W"
                            }
                        }
                        if(thresholdCritical) {
                            let regexC = new RegExp(thresholdCritical)
                            if( regexC.test(val) ) {
                                returnType= "C"
                            }
                        }
                    } catch(e) {
                        console.error(`regex failed`)
                    }
                }
            }
            return returnType
        }

        const formatValue = (val) => {
            let formattedVal=val

            let status = determineStatus(val)
            if(status==="N" && thresholdNormalLabel) {
                formattedVal = thresholdNormalLabel
            } else if(status==="W" && thresholdWarningLabel) {
                formattedVal = thresholdWarningLabel
            } else if(status==="C" && thresholdCriticalLabel) {
                formattedVal = thresholdCriticalLabel
            } else {
                //no labels provided so show value
                if(roundTo && !isNaN(roundTo) && roundTo > 0) {
                    formattedVal=Number(val).toFixed(roundTo)
                } else {
                    formattedVal=Math.round(Number(val))
                }
                formattedVal = isNaN(formattedVal) ? val : formattedVal //deal with nan's
            }
            return formattedVal
        }

        if(data) {
            let {current, history} = data

            let toolTipValueSuffix = valueSuffix ? " "+valueSuffix : ""
            let historyBlocks=[]
            for (let i=1; i <=24; i++) {
                if(history[history.length-i]) {
                    historyBlocks.push({status:determineStatus(history[history.length-i].value), value:formatValue(history[history.length-i].value)+toolTipValueSuffix, startTime: history[history.length-i].startTime, endTime: history[history.length-i].endTime})
                }
            }

            //extra info processor
            let info=null, infoTooltip=null
            if(customFeature) {
                //support for plugging in your own custom features here!
                switch(customFeature) {
                    case "example":
                        
                        let fdata=this.featureEXAMPLE(config,data)
                        info=(fdata && fdata.value) ? fdata.value : null
                        infoTooltip=(fdata && fdata.tooltip) ? fdata.tooltip : null
                        
                        break;
                    // case "somethingElse":
                    //     break;
                }
            }

            return <StatusBlock title={title} bigValue={formatValue(current)} bigValueLabel={valueLabel} bigValueSuffix={valueSuffix} status={determineStatus(current)} history={historyBlocks} info={info} infoTooltip={infoTooltip} link={link}/>
        } else {
            return <><StatusBlock title={title} /></>
        }
    }
}

function checkFieldName (field, results, isDebug) {
    let ret = field

    if (field == null) {
        if (Object.getOwnPropertyNames(results)[0] != null) {
            ret = Object.getOwnPropertyNames(results)[0]

            if (isDebug) {
                console.log(`No field name was specified, defaulting to first property name '${ret}'.`)
            }
        }
    }

    return ret
}
